lean/monad-stacks.lean

```diff
@@ -1,33 +1,250 @@
+
+#eval MyM.run 12 "hello" test -- "hello"
+#eval MyM.run 12 "hello" test2 -- "Number 12, String: hello"
+
+end Playground2
+/-!
+Nice! Everything just seems to work here, if you are interested in how
+Lean does this internally you can check out the infrastructure around:
+- `MonadReader`, `MonadReaderOf` for the `ReaderT` part
+- `MonadState`, `MonadStateOf` for the `StateT` part
+for most applications the idea that our `ReaderT` `StateT` monad stacks
+just combine nicely out of the box is perfectly fine though.
+
+### Lifting
+The next interesting question one might come up with is, can we just combine
+monad stacks nicely like this as well? What we are aiming for is that
+assuming we have a monad stack `S` and another one of our monad stacks `MyM`,
+that is built on top of `S`, we want to be able to execute computations inside of `S`
+just as easily inside of `MyM` as well. Let's see if the built-in stacks
+can do this already, as an example for `S` we will use the `IO` monad which is
+a monad stack under the hood as well:
+-/
+namespace Playground3
+
+abbrev Context := Nat
+abbrev State := String
+
+abbrev MyM  := ReaderT Context (StateT State IO)
+
+def test : MyM Unit := do
+  let old ← get
+  let new := s!"Number {←read}, String: {old}"
+  IO.println s!"Old was: {old}, New will be: {new}"
+  set new
+
+def MyM.run (context : Context) (state : State) (x : MyM α) : IO α :=
+  Prod.fst <$> StateT.run (ReaderT.run x context) state
+
+#eval MyM.run 12 "hello" test -- Old was: hello, New will be: Number 12, String: hello
+
+end Playground3
+/-!
+This also just worked without any additional effort on our side! The infrastructure
+responsible for this is `MonadLift` and `MonadLiftT` as well as some effort
+in the implementation of `do` to figure out when to lift. However you can again
+ignore this if you want, it just works out of the box most of the time.
+-/
+
+/-!
+## Error monads
+Next up we'll take a look at 2 monads you have most likely not heard of yet,
+both of which will show up in meta programming.
+
+First is `EStateM`, as it's name suggests it is very similar to `StateM` but
+with the fundamental difference that it can throw some sort of `E`rror,
+as such it is parameterized by 3 types: `EStateM Error State Value`.
+It is one of the monads that implement the `MonadExcept`/`MonadExceptOf`
+machinery which allows us to introduce a concept of exceptions into our
+monad stacks:
+-/
+namespace Playground4
+
+abbrev Context := Nat
+abbrev Error := String
+abbrev State := Nat
+
+abbrev MyM := ReaderT Nat (EStateM Error State)
+
+def alarm (x : String) : MyM Unit := do
+  throw s!"Alarm!: {x}"
+
+def main : MyM String := do
+  alarm "hi"
+  pure "test"
+
+def MyM.run (context : Context) (state : State) (x : MyM α) : Sum α Error :=
+  let res := EStateM.run (ReaderT.run x context) state
+  match res with
+  | .ok v _ => Sum.inl v
+  | .error e _ => Sum.inr e
+
+#eval MyM.run 12 13 main -- Sum.inr "Alarm!: hi"
```

Comment:
Again, I don't understand what happened. Is it because it didn't print "test". Maybe we should be more explicit